gbpservice/nfp/tests/unit/nfp/configurator/test_firewall.py

```diff
@@ -0,0 +1,887 @@
+            request_data = self.fo.fake_request_data_generic_bulk()
+            request_data1 = self.fo.fake_request_data_generic_bulk()
+        else:
+            request_data = self.fo.fake_request_data_generic_single()
+            request_data1 = self.fo.fake_request_data_generic_single()
+
+        if (method == 'CONFIGURE_ROUTES' or method == 'CLEAR_ROUTES'):
+            request_data1['config'][0]['resource'] = 'routes'
+            request_data['config'][0]['resource'] = 'routes'
+
+        with mock.patch.object(
+                    sc, 'new_event', return_value='foo') as mock_sc_event, \
+            mock.patch.object(sc, 'post_event') as mock_sc_rpc_event, \
+            mock.patch.object(rpc_mgr,
+                              '_get_service_agent_instance',
+                              return_value=agent):
+            if operation == 'create':
+                rpc_mgr.create_network_device_config(
+                                    self.fo.context, request_data)
+            elif operation == 'delete':
+                rpc_mgr.delete_network_device_config(
+                                    self.fo.context, request_data)
+
+            context = request_data1['config'][0]['kwargs']['context']
+            context.update({'notification_data': {}})
+            context.update({'resource': request_data1['config'][0]['resource']})
+            del request_data1['config'][0]['kwargs']['context']
+            kwargs = request_data1['config'][0]['kwargs']
+            if bulk:
+                sa_info_list = self.fo.fake_sa_info_list()
+                if operation == 'delete':
+                    sa_info_list[0]['method'] = 'clear_interfaces'
+                    sa_info_list[1]['method'] = 'clear_routes'
+                args_dict = {
+                         'sa_info_list': sa_info_list,
+                         'notification_data': {}
+                        }
+            else:
+                args_dict = {'context': context,
+                             'kwargs': kwargs}
+            mock_sc_event.assert_called_with(id=method, data=args_dict)
+            mock_sc_rpc_event.assert_called_with('foo')
+
+    def _test_fw_event_creation(self, operation):
+        sc, conf, rpc_mgr = self._get_ConfiguratorRpcManager_object()
+        agent, sc = self._get_FWaasRpcManager_object(conf, sc)
+        arg_dict = {'context': self.fo.context,
+                    'firewall': self.fo._fake_firewall_obj(),
+                    'host': self.fo.host}
+        method = {'CREATE_FIREWALL': 'create_network_service_config',
+                  'UPDATE_FIREWALL': 'update_network_service_config',
+                  'DELETE_FIREWALL': 'delete_network_service_config'}
+        request_data = self.fo.fake_request_data_fw()
+        with mock.patch.object(sc, 'new_event', return_value='foo') as (
+                                                        mock_sc_event), \
+            mock.patch.object(sc, 'post_event') as mock_sc_rpc_event, \
```

Comment:
All *aaS request are non-batch requests. They will not hit the event path in agent_base.py. We need to mock the agent method invocation right?